### DISCUSSION TITLE
Welcome to openopc2 Discussions!

### DISCUSSION THREAD

**Discussion by renzop** [iterativ/openopc2#4] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by supersuboy** [comment]
Hello!I have memory leak with opc.write. 
For example,my code is
————————————————————
from openopc2.da_client import OpcDaClient
opc_pc=''
opc = OpcDaClient()
opc_name_list = opc.servers(opc_pc)  # opcserver list
opc.connect('Matrikon.OPC.Simulation.1', opc_pc)
for i in range(0,10000000):
    opc.write(['T.PV',i])
    pass
————————————————————
Memory keeps increasing~~~ 

Thanks!

**Reply by renzop** [reply]
Thanks for your test,  can you deliver a bit more context.  Are you using the gateway server or com mode. Memory of wich process is increasing etc. Thanks

**Comment by ShionRiuka** [comment]
Hi all, 
I have a little issue:

My server and client sides are separated by a NAT
![image](https://user-images.githubusercontent.com/85865422/219056290-e40bce35-f8e9-4973-a554-be3501874fda.png)
Is there a specific function which combines the datagramm before sending it so I could twick it to work?

The problem here is with the handshake... the server sees two different IPs, where the packet was sent from and the client IP

**Reply by renzop** [reply]
note, that Openopc2 uses OPC DA (Opc classic) not OPC UA.

**Reply by ShionRiuka** [reply]
Yeah, I'm aware of it. It's a typo on the drowing.

**Reply by renzop** [reply]
Opc classic dos not connect over TCP IP but over com so the client and the server have to be on the same machine... NAT is not really the topic here. 

If you refere to Openopc2 gateway  ... (OPC tunneling if you look for a different name) then:

Can you test it first without the nat to check that everything is working ok? 
The openopc2 server / client (or gatway) uses pyro. Maybe you find the solution here 

https://pyro4.readthedocs.io/en/stable/tipstricks.html#pyro-behind-a-nat-router-firewall

Make sure you add the correct ports in your case (openopc2 uses 7766)

Anyway could you please modify your picture or clarify the question so that it is clearer.

**Reply by duduyoyo** [reply]
Classic OPC does use TCP! COM in classic OPC uses RPC as transportation layer, which works under TCP. 

If you don't want to deal with NAT, COM/DCOM, and gateway service installation etc., there is an alternative to let you retrieve classic OPC data easily. Take a look at this [one](https://github.com/duduyoyo/WebSocket4OPC).

**Comment by EulertronikTw** [comment]
Hi. Can you upload openopc2 excecutable version please for windows? I wish test it to get data from citect scada. Thanks in advance.

JP

**Reply by renzop** [reply]
You can download the executables in the repositories github page under
https://github.com/iterativ/openopc2/releases/tag/0.1.11

**Reply by EulertronikTw** [reply]
Thanks

**Comment by PardeepRassani** [comment]
> You can download the executables in the repositories github page under https://github.com/iterativ/openopc2/releases/tag/0.1.11

I am trying to install openopc2 on windows server 2019, I have download the three exe files, and tried to install OPC service by running `openopcservice install` and then `openopcservice start`  in the terminal, however it doesn't create OpenOPC directory in C drive, nor opc comand works in cmd. 

What am I missing please?

**Reply by danilNil** [reply]
@PardeepRassani why do you need an OpenOPC directory and what kind of command you expect to work in cmd? `openopcservice` is a windows service with no other commands except install, start, stop, restart.

**Comment by arunkumar790** [comment]
Hi, I am developing a OPC DA to UA wrapper using openopc2. It seems that the OpenOPC2 don't support to get Alarm and Event server data. Please correct me if I am wrong. Thanks for your support

**Comment by duduyoyo** [comment]
That is correct. The only way to support AE/HDA in Python is WebSocket4OPC. Google it.

**Reply by arunkumar790** [reply]
@duduyoyo Thanks for your response, checking WebSocket4OPC seems good.
